core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -1013,17 +1013,12 @@ object IO {
   final def bracket0[E, A, B](
     acquire: IO[E, A]
   )(release: (A, ExitResult[E, B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
-    Ref[Option[(A, ExitResult[E, B])]](None).flatMap { m =>
-      (for {
-        r <- (for {
-              a <- acquire
-              f <- use(a).fork
-              r <- f.observe
-              _ <- m.set(Some((a, r)))
-            } yield r).uninterruptibly
-        b <- r.fold(IO.now, IO.fail0, IO.terminate0)
-      } yield b).ensuring(m.get.flatMap(_.fold(unit) { case ((a, r)) => release(a, r) }))
-    }
+    Ref[Option[A]](None).flatMap { m =>
+      for {
+        f <- acquire.flatMap(a => m.set(Some(a)) *> use(a).fork).uninterruptibly
+        b <- f.join.ensuring(m.get.flatMap(_.fold(IO.unit)(a => f.observe.flatMap(r => release(a, r)))))
```

Comment:
The `ensuring` needs to go around the outermost block. Not around `join`, because if it's around `join`, then if `join` does not start executing (it's possible—the action can be interrupted right after the acquire block completes), then the finalizer will not execute. So move the `ensuring` to the outside of the `for` loop. This implies you need to put the fiber in the `Ref`, too.

Maybe something like:

```scala
    Ref[Option[(A, Either[Fiber[E, A], B])]](None).flatMap { m =>
      (for {
        f <- acquire.flatMap(a => use(a).fork.flatMap(f => m.set(Some(a -> Left(f)) *> IO.now(f)))).uninterruptibly
        b <- f.join
        _ <- m.update(_.map(t => (t._1, Right(b)))) // We have the `b`, so we don't need the fiber anymore
      } yield b).ensuring {
        m.get.flatMap(_.fold(IO.unit) {
          case (a, Left(fiber)) => fiber.observe.flatMap(release(a, _))
          case (a, Right(b))    => release(a, ExitResult.Completed(b))
        })
      }
    }
```

This is not quite correct. If you add `supervised` it won't really help.

We need something like `tryObserve` or `isDone` added to `Fiber`, so we can check to see if the fiber is done at the moment when the `ensuring` finalizer is being executed. If the fiber is not done, it means the main fiber is being interrupted, so we should interrupt the child fiber. If the fiber is done, then it means we should observe and use the observed exit result on the `release` action.

Reply:
😮 that's brilliant! Do we need the `Left`/`Right` ceremony? A `tryObserve` should be super fast, synchronous even. I've submitted a new version to let you see; will work on `tryObserve` now.

Reply:
I'm not sure, maybe you can eliminate it. The important thing is to pass the `b` to release if `use` succeeds.